### PR TITLE
Add MerkleTree

### PR DESCRIPTION
**Pull request by Recmo**

### CODE REVIEW

File: stark/src/merkle_tree/commitment.rs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+#[derive(Clone)]
+#[cfg_attr(feature = "std", derive(Debug))]
+pub struct Commitment {
+    depth: usize,
+    hash:  Hash,
+}
+
+impl Commitment {
+    pub fn from_depth_hash(depth: usize, hash: &Hash) -> Result<Self> {
```
Comment (z2trillion) [inline stark/src/merkle_tree/commitment.rs]: I found this useful discussion about whether to panic or to return a result here: https://www.reddit.com/r/rust/comments/9x17hn/when_should_a_library_panic_vs_return_result/

File: stark/src/merkle_tree/commitment.rs [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+    pub fn num_leaves(&self) -> usize {
+        1_usize << self.depth
+    }
+
+    pub fn hash(&self) -> &Hash {
+        &self.hash
+    }
+
+    /// Convert leaf indices to a sorted list of unique `Index`s.
```
Comment (z2trillion) [inline stark/src/merkle_tree/commitment.rs]: Can you add "And verifies that the indices are all valid"

File: stark/src/merkle_tree/index.rs [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+    #[quickcheck]
+    pub fn test_depth_offset_roundtrip(depth: usize, offset: usize) -> bool {
+        let depth = depth % Index::max_depth();
+        let offset = offset % Index::nodes_at_depth(depth);
+        let index = Index::from_depth_offset(depth, offset).unwrap();
+        index.depth() == depth && index.offset() == offset
+    }
+
+    #[quickcheck]
+    pub fn test_children(parent: Index) -> bool {
```
Comment (z2trillion) [inline stark/src/merkle_tree/index.rs]: Can you make this 
```
left.is_left() && right.is_right() && ...
```
?

File: stark/src/proofs.rs [hunk 4]
```diff
@@ -4,77 +4,95 @@ use crate::{
...
+// Clippy false positive
+#[allow(clippy::use_self)]
+impl VectorCommitment for Vec<MmapVec<FieldElement>> {
+    // TODO: Copy free implementation. Maybe have index as a leaf type.
+    type Leaf = Vec<U256>;
 
-    fn domain_size(&self) -> usize {
-        self.1.len() / self.0
+    fn len(&self) -> usize {
+        self.first().unwrap().len()
```
[the remark this answers is not in the document]
Comment (Recmo) [reply stark/src/proofs.rs]: Even better: If the outer vec is empty we have an empty commitment (which should be supported). In that case we can return zero.

[the remark this answers is not in the document]
Comment (Recmo) [reply stark/src/proofs.rs]: Actually, I have not covered the empty tree yet in a unit test! I'll add one.

File: stark/src/merkle_tree/commitment.rs (hunk 2 above)
Comment (Recmo) [reply stark/src/merkle_tree/commitment.rs]: I added it, but I think it goes without saying that a function validates it's input :)

File: stark/src/merkle_tree/index.rs (hunk 3 above)
Comment (Recmo) [reply stark/src/merkle_tree/index.rs]: I had that originally and didn't like it. When one of the subtests fails the backtrace will not tell you which one.

File: stark/src/merkle_tree/commitment.rs (hunk 2 above)
Comment (z2trillion) [reply stark/src/merkle_tree/commitment.rs]: IMO, from the function name, it's not clear why sorting indices could result in an error.

**Pull request merged**